add cells from paper http://jmlr.org/papers/volume17/14-203/14-203.pdf

gundramleifert opened the pull request:
These cells are tested in 2-dimensional RNNs and some of them yield to much better results. For 1-dimensional RNNs there are no good test - maybe one of you RWTH guys can substitute the LSTMs by these cells and check the performance. Especially for long sequences they maybe outperform the LSTM.

doetsch commented:
That was fast, thanks a lot! I will try it asap and will let you know about the results.

doetsch commented on NetworkRecurrentLayer.py:
```diff
@@ -130,6 +130,142 @@ def step(self, i_t, x_t, z_t, y_p, c_p, *other_args):
     # return: next outputs (# unit.n_act, y_t, c_t, ...)
     return (y_t * i_output, c_t * i_h + c_p * (1 - i_h)) + tuple(other_outputs)
 
+class Leaky(Unit):
```
Units need to be specified in all capital letters (the specification in the config file is case insensitive and will be capitalized to create the object). I.e. Leaky -> LEAKY, and maybe let's say LEAKYLSTM since the same idea could also be applied to GRUs for example.

doetsch commented on NetworkRecurrentLayer.py:
```diff
@@ -130,6 +130,142 @@ def step(self, i_t, x_t, z_t, y_p, c_p, *other_args):
+  This cell has 3 units instead of 4 like LSTM
+  """
+  def __init__(self, n_units, **kwargs):
+    super(LSTME, self).__init__(
```
Please provide the correct class name in the call of the super constructor (LSTME -> LEAKY)

doetsch commented on NetworkRecurrentLayer.py:
```diff
@@ -130,6 +130,142 @@ def step(self, i_t, x_t, z_t, y_p, c_p, *other_args):
+    return (y_t * i_output, c_t * i_h + c_p * (1 - i_h)) + tuple(other_outputs)
+
+
+class LeakyLP(Unit):
```
same here (capitalization required)

doetsch commented on NetworkRecurrentLayer.py:
```diff
@@ -130,6 +130,142 @@ def step(self, i_t, x_t, z_t, y_p, c_p, *other_args):
+  This cell has 4 units like the LSTM
+  """
+  def __init__(self, n_units, **kwargs):
+    super(LSTME, self).__init__(
```
same here (super constructor)

doetsch commented on NetworkRecurrentLayer.py:
```diff
@@ -130,6 +130,142 @@ def step(self, i_t, x_t, z_t, y_p, c_p, *other_args):
+  This cell has 4 units like the LSTM
+  """
+  def __init__(self, n_units, **kwargs):
+    super(LSTME, self).__init__(
```
... and here :)

gundramleifert commented:
I changed names to capitalized and changed the constructors. I wouldn't append "LSTM" to all structrues, I suggest LSTM, GRU and LEAKY are on the same hierarchy - they are all a fixed combination of basic components (a specific number of unit and update equations). I changed the name, but let's check first, if they work - then we can discuss about names ;-)
